### DISCUSSION TITLE
Do I need to handle frames in flight (queue.writeBuffer())?

### DISCUSSION THREAD

**Discussion by martinken** [gpuweb/gpuweb#2525] [Q&A] [answered]
Sorry if this is covered elsewhere. But in WebGPU, for best performance, do I need to double/triple buffer things like UBOs? For example in the WebGPU examples here

https://github.com/austinEng/webgpu-samples/blob/bbc01f17be0970bfdc312ea2a2ae823e9f436bd3/src/sample/rotatingCube/main.ts#L168

the camera matrix is updated every frame with a queue.writeBuffer(...) always to the same UBO (as opposed to having say 3 UBOs that round robin) So it made me wonder which approach I should be using for frequently changing UBOs. Just queue.writeBuffer() and let WebGPU deal with it, or setting up multiple UBOs and round robin them so that I'm always calling writeBuffer on one that isn't currently being used.

Thanks!

**Comment by kvark** [comment, marked as answer]
Using the same UBO is totally fine in this case. Doing double/triple-buffering is another strategy that you'd do without `writeBuffer`.

**Reply by kainino0x** [reply]
And even if you're not using writeBuffer (using mapAsync + copyBufferToBuffer instead), you wouldn't have to triple-buffer the uniform buffer because copyBufferToBuffer commands get enqueued in the same stream with draw calls. You'd only triple-buffer the mappable upload buffers (copyBufferToBuffer source buffers).

**Reply by martinken** [reply]
Makes sense, thanks so much @kvark @kainino0x for the answers.

**Comment by ghadeeras** [comment]
Not an answer, but related to the question, @toji put together a nice repo about WebGPU best practices. It already contains an article about using WebGPU buffers that I thought you (or any other passers-by) might find insightful: https://github.com/toji/webgpu-best-practices/blob/main/buffer-uploads.md

**Reply by martinken** [reply]
Great page @ghadeeras  That helped clarify the tradeoffs and the examples are very helpful. Sounds like the tradeoff for frequently updated buffers boils down to how much do you trust writeBuffer to be implemented well, as a good implementation of writeBuffer would make adding my own staging buffer ring extra complexity with little benefit.
